opengrok-web/src/main/java/org/opengrok/web/api/ApiTask.java

```diff
@@ -50,68 +55,145 @@
-        this.runnable = runnable;
+    public ApiTask(String path, Callable<Object> callable, Response.Status status) {
+        this(path, callable, status, null);
+    }
+
+   /**
+    * @param path request path (for identification)
+    * @param callable Callable object
+    * @param status request status to return after the runnable is done
+    * @param exceptionStatusMap map of {@link Exception} to {@link Response.Status}
+    */
+    public ApiTask(String path, Callable<Object> callable, Response.Status status,
+                   Map<Class<?>, Response.Status> exceptionStatusMap) {
+        this.callable = callable;
         this.uuid = UUID.randomUUID();
         this.responseStatus = status;
         this.path = path;
         this.state = ApiTaskState.INITIAL;
+        if (exceptionStatusMap != null) {
+            this.exceptionStatusMap.putAll(exceptionStatusMap);
+        }
     }
 
+    /**
+     * The UUID is randomly generated in the constructor.
+     * @return UUID
+     */
     public UUID getUuid() {
         return uuid;
     }
 
     /**
-     * @return response status
+     * @return response status to be used when the task was successfully completed
      */
-    public Response.Status getResponseStatus() {
+    Response.Status getResponseStatus() {
         return responseStatus;
     }
 
     /**
      * Set status as submitted.
      */
-    public void setSubmitted() {
+    void setSubmitted() {
         state = ApiTaskState.SUBMITTED;
     }
 
+    /**
+     * @return whether the API task successfully completed
+     */
     public boolean isCompleted() {
         return state.equals(ApiTaskState.COMPLETED);
     }
 
-    public void setCompleted() {
+    void setCompleted() {
         state = ApiTaskState.COMPLETED;
     }
 
+    /**
+     * @param future Future object used for tracking the progress of the API task
+     */
+    public void setFuture(Future<Object> future) {
+        this.future = future;
+    }
+
+    /**
+     * @return whether the task is finished (normally or with exception)
+     */
+    public boolean isDone() {
+        if (future != null) {
+            return future.isDone();
+        } else {
+            return false;
+        }
+    }
+
+    private Response.Status mapExceptionToStatus(ExecutionException exception) {
+        return exceptionStatusMap.getOrDefault(exception.getCause().getClass(), Response.Status.INTERNAL_SERVER_ERROR);
+    }
+
+    /**
+     * This method assumes that the API task is finished.
+     * @return response object corresponding to the state of the API task
+     * @throws IllegalStateException if the API task is not finished
+     */
+    public Response getResponse() {
+        // Avoid thread being blocked in future.get() below.
+        if (!isDone()) {
+            throw new IllegalStateException(String.format("task %s not yet done", this));
+        }
+
+        Object obj;
+        try {
+            obj = future.get();
```

Comment:
`future` can be `null` here.

Reply:
I think it is safe given the `isDone()` check above.